man/pinspect.1.rst
==========
 pinspect
==========

----------------------------------------------------
arbitrary reporting and inspection of configuration
----------------------------------------------------

:Date:   2011-09-29
:Manual section: 1
:Manual group: pkgcore

SYNOPSIS
========

  pinspect subcommand [options]

DESCRIPTION
===========

pinspect is for querying pkgcore for reports, and general non package querying
information.  For example, to find eapi usage for a repository, the eapi_usage
subcommand exists; for portageq like querying, a portageq subcommand exists.

OPTIONS
=======

.. pkgcore_script_options:: pkgcore.scripts.pinspect.commandline_commands


NEWS
Pkgcore Release Notes
=====================

See ChangeLog for full commit logs; this is summarized/major changes.


pkgcore 0.6.4: Jun 5th, 2011

* intercept and suppress exceptions from triggers unless the trigger
  explicitly disables it.

* work around libmagic python bindings sucking and not always being
  able to be used.

* fix 'default' support for src_install for EAPI=4.


pkgcore 0.6.3: May 30th, 2011

* support for /etc/portage/make.profile; Please Do Not Use it, while
  pkgcore is forced to support it, usage of it breaks most tools and is
  bluntly lock-in (no reason to move it- it's the same, been in the same
  place for a decade now).  Duly warned.

* misc env/bug fixes for EMERGE_FROM to ensure compatibility.

* deploy eselect support via pinspect portageq

* added man page for pinspect

* added pmaint env-update

* expose /usr/local/* through PATH for ebulids.


pkgcore 0.6.2: May 27th, 2011

* for EAPI<4, expose MERGE_TYPE info via EMERGE_FROM; do this for compatibility
  with non-spec compliant ebuilds, and eclasses like linux-mod.  This restores
  in particular, binpkg support for kernel packages.  Thanks to Brian De Wolf
  for info leading to tracking this down.

* add support for stacking /etc/portage/make.conf on top of /etc/make.conf.

* add incrementalism between make.globals and make.conf to match changes
  in portage configuration parsing.  This fixes the common "I tried pkgcore
  and everything was license masked".  Breakage there owes to portage
  changing make.globals; can't do much about it unfortunately.  Thanks to
  Brian De Wolf for info leading to tracking this down.

* prefer 0755 permissions for binpkg package dir.

* pinspect pkgset learned --all option, to display all pkgsets it knows.


pkgcore 0.6.1: May 22nd, 2011

* fix for "or_node.blocks" AttributeError, and related resolution
  miscalculations.

* fix exit code return for ebuild helpers throwing warnings for <EAPI4

* fix typo in FEATURES=buildsyspkg, and FEATURES=buildpkg

* check to ensure pkgdir exists; if possible, create it, else turn off binpkg
  features.


pkgcore 0.6: April 24th, 2011

* Due to crazy work hours and moves, this release is fairly large, and frankly
  repeatedly delayed.  Future ones will be far more fine grained moving forward.

* Fix python2.7 incompatiblity in pkgcore.ebuild.misc

* It's suggested that folks use bash 4.1, primarily for regen
  speed reasons- it is not required however.

* bash spawning now enforce --norc and --noprofile in full.

* RESTRICT is now properly use evaluated.

* pkgcore.restrictions.values.ContainmentMatch is deprecated in favor of
  ContainmentMatch2.  Update your code- by pkgcore 0.7, ContainmentMatch
  will become ContainmentMatch2, and a shim will be left in place.

* introduction of EAPI objects (pkgcore.ebuild.eapi) for controlling/defining
  new eapi's, capabilities, etc.

* pmaint regen is now cumulatively ~23x faster then the previous release.
  This is via restoration of original metadata regeneration speeds, and
  via enabling an eclass preloading optimization.  No impact on metadata-
  just far faster regeneration.

* Roughly a 15x speedup in general metadata regeneration; basically rework
  a fix that was added to to 0.5.11 (dealing with portage induced
  breakage in env loadup from their declare usage).

* filter-env regex backend now uses python's re always; previously
  if the extension was active it would use posix regex.

  This resolves occasional odd failures when running native filter-env.

* fix a truncation error in suffix version parsing resulting in
  _p2010081516093 being less than _p2009072801401 .

* pkgcore.ebuild.restricts now contains some generally useful
  building block restrictions for any api consumers

* full rewrite of EAPI helpers adding better error info, saner code,
  double checked against PMS and portage/paludis to ensure no oddities.

* fix to buildpkg/pristine-binpkg saving.  If you're looking for
  something to contribute to pkgcore wise, tests for this would be
  appreciated.

* write support for DEFINED_PHASES.

* bashrc hooks now run from ${S} or ${WORKDIR}, depending on
  PMS rules for that phase.

* match the other PM's for econf; update ${WORKDIR} instances of
  config.{sub,guess} from /usr/share/gnuconfig.

* added protection against bad environment dumps from other PMs for T
  during env restoration.

* removed RESTRICT=autoconfig support.

* fix compatibility regression introduced in file-5.05 involving MAGIC_NONE.

* handle keyboard interrupts better during compilation; no longer display
  die tracebacks if the user intentionally stopped the compilation.

* duplicate a portage workaround for emacs ebuild; specifically don't
  regenerate infodir if the ebuild placed a .keepinfodir in the directory.
  gentoo bug #257260.

* add workaround to disable unzip during unpack going interactive during
  a failure; gentoo bug #336285.

* fixed traceback during displaying a summary for 'pinspect eapi_usage'

* add EAPI limitation to all portageq invocations, and support USE dep
  usage with has_version and friends.

* handle portage's new interpretation of the sync retries variable for portage
  configuration.

* pinspect distfiles_usage was added; this is primarily useful for getting
  a repository level view of what the distfiles requirements are, what takes
  what percentile of unique space, etc.

* FEATURES=allow-missing-manifests ; does exactly as it sounds, not advised to
  use unless you know what you're doing.

* ospkg's fork of pkgcore has been folded in; FEATURES=save-deb is the primary
  addition.

* extended atom syntax now allows '*' to be used w/in a string- for example
  dev-*kde, *dev-*k*de*, etc.  This syntax is usable in user configs, and
  from the commandline.

* new FEATURES=fixlafiles is on by default; basically folds
  dev-util/lafilefixer functionality directly into the merger.
  Note this version drops comments- it's about a 66% reduction in .la system

pkgcore/scripts/pinspect.py
# Copyright: 2009-2011 Brian Harring <ferringb@gmail.com>
# License: GPL2/BSD

__all__ = ("pkgsets_data", "histo_data", "eapi_usage_data", "license_usage_data",
    "mirror_usage_data", "eclass_usage_data", "mirror_usage_data"
)

from pkgcore.util.commandline import OptionParser
from pkgcore.ebuild import portageq
import os
from snakeoil.demandload import demandload

demandload(globals(),
    'snakeoil.lists:iflatten_instance,unstable_unique',
    'snakeoil.mappings:defaultdict',
    'pkgcore:fetch',
    'pkgcore.restrictions:packages',
    'itertools:groupby,islice',
    'operator:attrgetter,itemgetter'
)

commandline_commands = {}

class pkgsets_data(OptionParser):

    #enable_domain_options = True
    description = 'inspect pkgsets available in configuration'
    usage = ('%prog pkgsets [sets-to-examine] ; if no sets specified, '
        'list the known sets')

    def _register_options(self):
        self.add_option("--all", action='store_true', default=False,
            help="display info on all pkgsets")

    def _check_values(self, values, args):
        values.pkgsets = tuple(args)
        if values.all and values.pkgsets:
            self.error("--all cannot be specified along w/ specific pkgsets to inspect")
        return values, ()

    def run(self, opts, out, err):
        if opts.all:
            opts.pkgsets = sorted(opts.config.pkgset)
        if not opts.pkgsets:
            out.write(out.bold, 'available pkgset: ', out.reset,
                ', '.join(repr(x) for x in
                   sorted(opts.config.pkgset)))
            return 0
        missing = False
        for position, set_name in enumerate(opts.pkgsets):
            pkgset = opts.get_pkgset(None, set_name)
            if pkgset is None:
                missing = True
                if position:
                    out.write()
                out.write(out.bold, 'pkgset ', repr(set_name), out.reset,
                    "isn't defined, skipping.")
            else:
                if position:
                    out.write()
                out.write(out.bold, 'pkgset ', repr(set_name), out.reset, ':')
                out.first_prefix.append('  ')
                for restrict in sorted(pkgset):
                    out.write(restrict)
                out.first_prefix.pop()
        if not missing:
            return 0
        return 1

commandline_commands['pkgset'] = pkgsets_data


def print_simple_histogram(data, out, format, total, sort_by_key=False,
    first=None, last=None):
    # do the division up front...

    total = float(total) / 100

    if sort_by_key:
        data = sorted(data.iteritems(), key=itemgetter(0))
    else:
        data = sorted(data.iteritems(), key=itemgetter(1), reverse=True)

    if first:
        data = islice(data, 0, first)
    elif last:
        data = list(data)[-last:]

    for key, val in data:
        out.write(format % {'key':str(key), 'val':val,
            'percent':"%2.2f%%" % (val/total,)})


class histo_data(OptionParser):

    per_repo_summary = None
    allow_no_detail = False

    def _register_options(self):
        self.add_option("--no-final-summary", action='store_true', default=False,
            help="disable outputting a summary of data across all repos")

        if self.per_repo_summary:
            self.add_option("--no-repo-summary", action='store_true', default=False,
                help="disable outputting repo summaries")

        if self.allow_no_detail:
            self.add_option("--no-detail", action='store_true', default=False,
                help="disable outputting a detail view of all repos")

        self.add_option("--sort-by-name", action='store_true', default=False,
            help="sort output by name, rather then by frequency")

        self.add_option("--first", action="store", type='int', default=0,
            help="show only the first N detail items")

        self.add_option("--last", action="store", type='int', default=0,
            help="show only the last N detail items")

    def _check_values(self, opts, args):
        repo_conf = opts.config.repo
        if args:
            opts.repos = []
            for repo_name in args:
                if not repo_name in repo_conf:
                    self.error("no repository named %r" % (repo_name,))
                opts.repos.append((repo_name, repo_conf[repo_name]))
        else:
            opts.repos = sorted(repo_conf.items(), key=lambda x:x[0])

        if not self.allow_no_detail:
            opts.no_detail = False

        if not self.per_repo_summary:
            opts.repo_summary = False

        if opts.no_detail and opts.no_repo_summary and opts.no_final_summary:
            s = '--no-final-summary '
            if self.allow_no_detail:
                s += 'and --no-detail '
            if self.per_repo_summary:
                s += 'and --no-repo-summary '
            self.error("%s cannot be used together; pick just one" % (s,))

        if opts.last and opts.first:
            self.error("--first and --last cannot be used together; use just one")

        if not self.per_repo_summary:
            opts.no_repo_summary = True

        if not self.allow_no_detail:
            self.no_detail = False

        return opts, ()

    def get_data(self, repo, options):
        raise NotImplementedError()

    def transform_data_to_detail(self, data):
        return data

    def transform_data_to_summary(self, data):
        return data

    def run(self, opts, out, err):
        global_stats = {}
        position = 0
        total_pkgs = 0
        for repo_name, repo in opts.repos:
            if position:
                out.write()
            position += 1
            out.write(out.bold, "repository", out.reset, ' ',
                repr(repo_name), ':')
            data, repo_total = self.get_data(repo, opts)
            detail_data = self.transform_data_to_detail(data)
            if not opts.no_detail:
                out.first_prefix.append("  ")
                if not data:
                    out.write("no pkgs found")
                else:
                    print_simple_histogram(detail_data,
                        out, self.per_repo_format, repo_total,
                        sort_by_key=opts.sort_by_name,
                        first=opts.first, last=opts.last)
                out.first_prefix.pop()
            for key, val in detail_data.iteritems():
                global_stats.setdefault(key, 0)
                global_stats[key] += val
            total_pkgs += repo_total

            if opts.no_repo_summary:
                continue
            out.write(out.bold, 'summary', out.reset, ': ',
                self.per_repo_summary %
                self.transform_data_to_summary(data))

        if position > 1 and not opts.no_final_summary:
            out.write()
            out.write(out.bold, 'summary', out.reset, ':')
            out.first_prefix.append('  ')
            print_simple_histogram(global_stats, out, self.summary_format,
                total_pkgs, sort_by_key=opts.sort_by_name)
            out.first_prefix.pop()
        return 0


class eapi_usage_data(histo_data):

    #enable_domain_options = True
    description = 'get a breakdown of eapi usage for target repositories'
    usage = ('%prog eapi_usage [repositories to look at] ; if no repositories '
        'specified the default is to scan all')

    per_repo_format = ("eapi: %(key)r %(val)s pkgs found, %(percent)s of the "
        "repository")

    summary_format = ("eapi: %(key)r %(val)s pkgs found, %(percent)s of all "
        "repositories")

    def get_data(self, repo, options):
        eapis = {}
        pos = 0
        for pos, pkg in enumerate(repo):
            eapis.setdefault(pkg.eapi, 0)
            eapis[pkg.eapi] += 1
        return eapis, pos + 1

commandline_commands['eapi_usage'] = eapi_usage_data


class license_usage_data(histo_data):

    description = 'get a breakdown of license usage for target repositories'
    usage = ('%prog license_usage [repositories to look at] ; if no repositories '
        'specified the default is to scan all')

    per_repo_format = ("license: %(key)r %(val)s pkgs found, %(percent)s of the "
        "repository")

    summary_format = ("license: %(key)r %(val)s pkgs found, %(percent)s of all "
        "repositories")

    def get_data(self, repo, options):
        data = {}
        pos = 0
        for pos, pkg in enumerate(repo):
            for license in unstable_unique(iflatten_instance(pkg.license)):
                data.setdefault(license, 0)
                data[license] += 1
        return data, pos + 1

commandline_commands['license_usage'] = license_usage_data


class eclass_usage_data(histo_data):

    description = 'get a breakdown of eclass usage for target repositories'
    usage = ('%prog eclass_usage [repositories to look at] ; if no '
        'repositories are specified it defaults to scanning all')

    per_repo_format = ("eclass: %(key)r %(val)s pkgs found, %(percent)s of the "
        "repository")

    summary_format = ("eclass: %(key)r %(val)s pkgs found, %(percent)s of all "
        "repositories")

    def get_data(self, repo, options):
        pos, data = 0, {}
        for pos, pkg in enumerate(repo):
            for eclass in getattr(pkg, 'data', {}).get("_eclasses_", {}).keys():
                data.setdefault(eclass, 0)
                data[eclass] += 1
        return data, pos + 1

commandline_commands['eclass_usage'] = eclass_usage_data


class mirror_usage_data(histo_data):

    description = 'get a breakdown of mirror usage for target repositories'
    usage = ('%prog mirror_usage [repositories to look at] ; if no '
        'repositories are specified it defaults to scanning all')

    per_repo_format = ("mirror: %(key)r %(val)s pkgs found, %(percent)s of the "
        "repository")

    summary_format = ("mirror: %(key)r %(val)s pkgs found, %(percent)s of all "
        "repositories")

    def get_data(self, repo, options):
        data = {}
        for pos, pkg in enumerate(repo):
            for fetchable in iflatten_instance(pkg.fetchables,
                fetch.fetchable):
                for mirror in fetchable.uri.visit_mirrors(
                    treat_default_as_mirror=False):
                    if isinstance(mirror, tuple):
                        mirror = mirror[0]
                    data.setdefault(mirror.mirror_name, 0)
                    data[mirror.mirror_name] += 1
        return data, pos + 1

commandline_commands['mirror_usage'] = mirror_usage_data


class distfiles_usage_data(histo_data):

    description = 'get a breakdown of total distfiles for target repositories'
    usage = ('%prog mirror_usage [repositories to look at] ; if no '
        'repositories are specified it defaults to scanning all')

    per_repo_format = ("package: %(key)r %(val)s bytes, referencing %(percent)s of the "
        "unique total")

    per_repo_summary = ("unique total %(total)i bytes, sharing %(shared)i bytes")

    summary_format = ("package: %(key)r %(val)s pkgs found, %(percent)s of all "
        "repositories")

    allow_no_detail = True

    def _register_options(self):
        histo_data._register_options(self)
        self.add_option("--include-nonmirrored", action='store_true', default=False,
            help="if set, nonmirrored  distfiles will be included in the total")
        self.add_option("--include-restricted", action='store_true', default=False,
            help="if set, fetch restricted distfiles will be included in the total")

    def get_data(self, repo, options):
        owners = defaultdict(set)
        iterable = repo.itermatch(packages.AlwaysTrue, sorter=sorted)
        items = {}
        for key, subiter in groupby(iterable, attrgetter("key")):
            for pkg in subiter:
                if not options.include_restricted and 'fetch' in pkg.restrict:
                    continue
                if not options.include_nonmirrored and 'mirror' in pkg.restrict:
                    continue
                for fetchable in iflatten_instance(pkg.fetchables, fetch.fetchable):
                    owners[fetchable.filename].add(key)
                    items[fetchable.filename] = fetchable.chksums.get("size", 0)

        data = defaultdict(lambda:0)
        for filename, keys in owners.iteritems():
            for key in keys:
                data[key] += items[filename]
        unique = sum(items.itervalues())
        shared = sum(items[k] for (k,v) in owners.iteritems() if len(v) > 1)
        return (data, {"total":unique, "shared":shared}), unique

    def transform_data_to_detail(self, data):
        return data[0]

    def transform_data_to_summary(self, data):
        return data[1]

commandline_commands['distfiles_usage'] = distfiles_usage_data
commandline_commands['portageq'] = portageq.commandline_commands
